arxiv/metadata/metacheck.py

```diff
@@ -0,0 +1,687 @@
+    """
+    if re.search(s1, s2, re.IGNORECASE) \
+       and not re.search("[$].*"+s1+".*[$]", s2, re.IGNORECASE):
+        return True
+    else:
+        return False
+       
+
+############################################################
+# This is the main function
+
+def check(metadata:Dict[FieldName,str]): 
+    result = {} 
+    for (k,v) in metadata.items():
+        if k == TITLE:
+            result[k] = check_title(v)
+        elif k == AUTHORS:
+            result[k] = check_authors(v)
+        elif k == ABSTRACT:
+            result[k] = check_abstract(v)
+        elif k == SOURCE_TYPE:
+            result[k] = check_source_type(v)
+        elif k == COMMENTS:
+            result[k] = check_comments(v)
+        elif k == REPORT_NUM:
+            result[k] = check_report_num(v)
+        elif k == JOURNAL_REF:
+            result[k] = check_journal_ref(v)
+        elif k == DOI:
+            result[k] = check_doi(v)
+        else:
+            raise UnknownMetadataException(f"Unknown metadata field {k}: {v}")
+    #
+    return result
+
+def looks_like_all_caps(s: str) -> bool:
+    """
+    Returns true if s appears to have all/excessive capitals. The
+    exact threshold is heuristic based on looking at submissions over
+    March 2010.
+    """
+    num_caps = s.upper()
+    num_caps = sum([ c.isupper() for c in s])
+    num_lower = sum([ c.islower() for c in s])
+    if num_caps > num_lower * 2 + 7:
+        return True
+    else:
+        return False
+
+
+def long_word_caps(s: str) -> bool:
+    """
+    Returns true (was: a list of long words) if s appears to have
+    two or more words which are not in our list
+    KNOWN_WORDS_IN_ALL_CAPS.
+
+    """
+    num_matches = 0
+    for word in s.split():
+        if len(word) >= 6 and \
+           word.upper() == word and \
+           word.lower() != word and \
+           (word not in KNOWN_WORDS_IN_ALL_CAPS):
+            num_matches += 1
+        #
+    #
+    if num_matches > 1:
+        return True
+    else:
+        return False
+    #
+
+def smart_starts_with_lowercase_check(s: str) -> bool:
+    """
+    Detect titles which start with a lower case char, eg "bad title"
+    but do not reject "eSpeak: A New World"
+    """
+
+    if len(s) < 1:
+        return False
+    if re.match( "[a-z]", s):
+        if re.match( "[a-z][A-Z][a-zA-Z]*: [A-Z]", s):
+            return False
+        else:
+            return True
+    else:
+        return False
+    #
+
+def all_brackets_balanced(s: str) -> bool:
+    pending_brackets = []
+    for c in s:
+        if c == "(":
+            pending_brackets.append(")")
+        elif c == "[":
+            pending_brackets.append("]")
+        elif c == "{":
+            pending_brackets.append("}")
+        elif len(pending_brackets) > 0 and c == pending_brackets[-1]:
+            pending_brackets = pending_brackets[:-1]
+        elif c in ")}]":
+            return False
+    #

```

Comment:
remove extra comment hash

Reply:
I put comments in to denote "end if", which Python syntax does not support.  Otherwise, some tools break.

Reply:
:+1: